### ISSUE TITLE
[DX] Reducing direct dependencies on the Router

### ISSUE THREAD

**Issue by iltar** [symfony/symfony#15319]
I've noticed one common piece of code when I have to write a redirect: 

```php
return new RedirectResponse($this->router->generate('app.example', [...]));
```

What I would like to write is something similar to:

```php
return new RoutedRedirectResponse('app.example', [...]);
```
The actual route would be generated at a later stage and use the default router.

**Comment by javiereguiluz** [comment]
The name of the new class may be `RedirectToRouteResponse()` to be consistent with `redirectToRoute()` helper.

**Comment by javiereguiluz** [comment]
I lnow that it will never be accepted, but I'd like to see the following:

```php
// the "URL" starts with '@' -> consider it a route instead
return new RedirectResponse('@app.example', [...]);
```

**Comment by iltar** [comment]
That would make it very consistent with Symfony 1, but requires an argument change in the redirect response itself.

**Comment by GromNaN** [comment]
This could be implemented with a response listener that converts the route + parameters to an URL.

**Comment by stof** [comment]
The issue with this is that your ``RoutedRedirectResponse`` cannot be a Response object. It stores information needed to generate a Response at a later time.

If you want something like that, the way to go would be to use a *different* object and then have a kernel.view listener converting it into a Response. and btw, this already exists in a bundle: https://github.com/QafooLabs/QafooLabsNoFrameworkBundle#redirect-route

**Comment by fabpot** [comment]
I'm :-1: on this one as I was :-1: on the `redirectToRoute`, which I found unnecessary and against what people refer to "best practices".

TLDR: I really prefer the explicit way and having only one way to do things.

**Comment by iltar** [comment]
For one of my Applications it would mean roughly 50 direct dependencies less on the router. Another solution could be a Response factory which could be reused in the `Controller` that Symfony provides, as well as wrapping around the  template engine to cover the basic render methods in there.

It might be more work than worth it though.

**Comment by iltar** [comment]
What about providing a `RedirectResponseFactory`?
```php
    // $contract = new Contract();
    return $this->factory->generateRedirect('app.contract.view', ['contract' => $contract);
```

This could also provide a proper abstraction layer to convert a contract to the parameter defined in the route. Right now I have to write `['contract' => $contract->getId()]`. In case my Product Owner decides it has to be changed to a slug, I will have to update every single url generation and update it, where as with an abstraction layer I can just change the converter.

**Comment by WouterJ** [comment]
You can always create your own factory.

I don't see the benefit of the factory, as you will still have a dependency. You just changed it from a dep on router to a dep on the redirect response factory.

**Comment by iltar** [comment]
Yes I can, but before designing something like this I want to see if it fits in Symfony. Yes I will still have a dependency, but I will no longer have to create the `RedirectResponse` in the controller itself and I will have solved my issue of having to "duplicate" the `->getId()` being unable to easily change it by implementing a reversed parameter converter.

**Comment by iltar** [comment]
I've solved my issue in https://github.com/iltar/http-bundle, for those who are interested. All I have to do is build a redirect response factory and my issue is solved.

Thanks for the feedback!
